### PR TITLE
Statistics framework design simplifications

### PR DESCRIPTION
**Pull request by iulianbarbu**
# Reason for This PR

Incremental improvement over the statistics framework design.

## Description of Changes

* Introduced the `MetadataProvider`, a component able to abstract statistics and measurements definitions.
* Updated existing tests to use this component.
* The direct interface for statistics and measurements definitions is still available for tests where a `MetadataProvider` is simply too much (e.g test_network_latency).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
~~- [ ] Any newly added `unsafe` code is properly documented.~~
~~- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.~~
~~- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.~~
~~- [ ] All added/changed functionality is tested.~~

### CODE REVIEW

File: tests/integration_tests/performance/test_network_latency.py [hunk 1]
```diff
@@ -32,33 +32,39 @@
...
-                                              function.Percentile90,
-                                              function.Percentile99],
-                                             pass_criteria())
-    stats_defs.append(consumer.StatisticDef(PKT_LOSS_STAT_KEY,
-                                            PKT_LOSS,
-                                            function.GetFirstObservation))
+    stats_defs = types.StatisticDef.defaults(
+        LATENCY,
+        # We are going to consume multiple statistics
+        [function.ValuePlaceholder(function.Avg.__name__),
```
Comment (alindima) [inline tests/integration_tests/performance/test_network_latency.py:56]: seems rather verbose. the `name()` method seems to still be available.
Moreover, tying this test to the `function.*` functions can be misleading. The only thing it uses from that classes are the names. I'd say just provide local names

File: tests/integration_tests/performance/test_block_performance.py [hunk 2]
```diff
@@ -70,146 +58,39 @@ def __init__(self, cpu_model_name):
...
-        """Return the target value corresponding to the key."""
-        return self._baselines.get(key)["target"]
-
-    def delta(self, key: str) -> Number:
-        """Return the delta value corresponding to the key."""
-        return self.target(key) * self._baselines.get(key)[
-            "delta_percentage"] / 100
-
-
-def cpu_utilization_measurements():
```
Comment (alindima) [inline tests/integration_tests/performance/test_block_performance.py:83]: Loving the fact that we can move these definitions from this test file 👍🏻

File: tests/integration_tests/performance/configs/vsock_throughput_test_config.json [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+  "statistics": {
+    "throughput": [{
+      "name": "total",
+      "function": "Sum",
+      "criteria": "EqualWith"
+    }],
+    "duration": [{
+      "function": "Avg"
+    }],
+    "cpu_utilization_vmm": [{
```
Comment (alindima) [inline tests/integration_tests/performance/configs/vsock_throughput_test_config.json:43]: These should also have criteria

File: tests/framework/statistics/function.py [hunk 4]
```diff
@@ -3,155 +3,162 @@
...
+    @property
+    def name(self) -> str:
+        """Return the a name identifier for the class."""
+        return self._name
 
 
 # pylint: disable=R0903
-class GetFirstObservation(StatisticFunction):
-    """A function which return the first observation."""
+class ValuePlaceholder(Function):
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/framework/statistics/function.py:41]: I had this in picture once but I think that using interchangeably measurements and statistics for the corner case you describe will bring more confusion to the table. Statistics and measurements have defining roles in this framework and helps grasping the resources needed by a successful statistical exercise. Statistics refer to insights over data, while the measurements refer to the numerical quantitation of the attributes of an object. 

Putting all of this together will lead us to insights (statistics) over the attributes (measurement name) to numerical quantitation (measurement unit) of an object (network, block, vsock). I find every piece defined previously very important for concrete expectations throughout the test design and format of the test result.

Regarding the fact we have to run 1 iteration for getting statistics data, this is not necessary the case (it is for existing tests, but we need to keep room for other scenarios) - e.g a tool produces for 10 iterations measurement data and also in the 10th iteration tells you the average CPU utilization for the previous 10 iterations.

Let me know in case your questions were not answered.

File: tests/integration_tests/performance/test_network_latency.py (hunk 1 above)
Comment (iulianbarbu) [reply tests/integration_tests/performance/test_network_latency.py]: Done. Please take a look.

The `name()` method is an object one and it is mainly relevant post definition, when for a specific statistic definition we want to get its assigned name. Previously, the method was thought of as a class method, when the functions were lazily instantiated. I've changed this aspect to remove redundant information and for separation of concerns - the functions are the sole owner of their processing identification and this identification is used by statistics as their name.

File: tests/integration_tests/style/test_licenses.py [hunk 5]
```diff
@@ -2,9 +2,10 @@
 # SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/integration_tests/style/test_licenses.py:2]: Done.

**Review by iulianbarbu — COMMENTED** [review]
@alindima , finished to integrate your feedback. Take a look when you have time: https://github.com/firecracker-microvm/firecracker/pull/2424/commits/e364ad88c339e26378724752306d2c3a72aec9ec.

File: tests/integration_tests/style/test_licenses.py [hunk 6]
```diff
@@ -2,9 +2,10 @@
...
 """Tests checking against the existence of licenses in each file."""
 
+from datetime import datetime
 import framework.utils as utils
 
-AMAZON_COPYRIGHT_YEARS = (2018, 2019, 2020)
+AMAZON_COPYRIGHT_YEARS = range(2018, datetime.now().year + 1)
```
Comment (gc-plp) [inline tests/integration_tests/style/test_licenses.py:8]: Isn't this addressed in https://github.com/firecracker-microvm/firecracker/pull/2423/commits/fe26cb2c90a1923dce645939094ee3fcffc4380e ?

Comment (alindima) [reply tests/integration_tests/style/test_licenses.py:8]: it is. It's just a race for the rebase :))

File: tests/integration_tests/performance/test_vsock_throughput.py [hunk 7]
```diff
@@ -8,85 +8,77 @@
...
-    """Define the produced measurements for VSOCK workloads."""
-    return [types.MeasurementDef(test_cfg.THROUGHPUT,
-                                 test_cfg.THROUGHPUT_UNIT),
-            types.MeasurementDef(test_cfg.DURATION, test_cfg.DURATION_UNIT),
-            types.MeasurementDef.cpu_utilization_vmm(),
-            types.MeasurementDef.cpu_utilization_vcpus_total()]
 
+# Measurements units.
+CONFIG_RAW_FILE = os.path.join(
+    os.path.dirname(__file__),
```
Comment (gc-plp) [inline tests/integration_tests/performance/test_vsock_throughput.py:28]: What about defining a base folder that contains the config files and just do a one liner?

```
# snip
CFG_LOC = Path("/well/known/location")
# end of snip

CONFIG = json.load(open(CFG_LOC / "vsock_throughput_test_config.json"))
```

Comment (iulianbarbu) [reply tests/integration_tests/performance/test_vsock_throughput.py]: Done.

File: tests/integration_tests/performance/configs/vsock_throughput_test_config.json (hunk 3 above)
Comment (iulianbarbu) [reply tests/integration_tests/performance/configs/vsock_throughput_test_config.json:43]: Omitted this one, please take a look here: https://github.com/firecracker-microvm/firecracker/pull/2424/commits/97dbe9aef8c1c2301de0abfe43aab74897df55b9.

**Comment by iulianbarbu** [comment]
@alindima , @gc-plp , I've done some styling fixes and also other fixes necessary due to tests not being able to start. I've updated the commits correspondent to your feedback. Let me know in case you find something else.

**Comment by gc-plp** [comment]
I think that the changes do bring some value to the code, so let's go ahead with merging it.

**Comment by iulianbarbu** [comment]
@gc-plp , @alindima , I've rebased the commits for final review. Let me know in case there are other comments.

File: tests/integration_tests/performance/test_vsock_throughput.py [hunk 8]
```diff
@@ -341,8 +310,7 @@ def iperf_workload(context):
 
     st_core = core.Core(name="vsock_throughput",
                         iterations=1,
-                        check_criteria=file_dumper is None,
```
Comment (alindima) [inline tests/integration_tests/performance/test_vsock_throughput.py:344]: Please move this check to the `run_exercise()` call

**Review by alindima — COMMENTED** [review]
Looks good, mostly nits and style comments

File: tests/framework/statistics/criteria.py [hunk 9]
```diff
@@ -26,65 +26,120 @@ def msg(self, msg):
...
 
-    def __init__(self, target: Number):
+    def __init__(self, baseline: dict):
         """Initialize the criteria."""
-        super().__init__("LowerThan", target)
+        super().__init__("LowerThan", baseline)
 
     def check(self, actual):
         """Compare the target and the actual."""
         fail_msg = self.name + f" failed. Target: '{self.target} " \
```
Comment (alindima) [inline tests/framework/statistics/criteria.py:105]: I'd say to also have this `fail_msg` as a property method on the base class, which can be overriden on a case-by-case basis (for the `EqualWith` operator), to minimise duplication

File: tests/framework/statistics/__init__.py [hunk 10]
```diff
@@ -1,12 +1,316 @@
...
 
-"""Single threaded producer/consumer for statistics gathering."""
+"""
+Single threaded producer/consumer for statistics gathering.
+
+The purpose of this module is to provide primitives for statistics exercises
+which need a common framework that sets expectations in terms of tests
+design and results.
+
+The main components of the module consist from: `Core`, `Producer` and
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/framework/statistics/__init__.py:11]: It is already done. I think you've reviewed a previous state of the same phrase.

File: tests/framework/statistics/__init__.py [hunk 11]
```diff
@@ -1,12 +1,316 @@
...
+pass criteria for measurements statistics.
+```
+CONFIG = {
+    "measurements": {
+        # This is a map where keys represent the exercise id while the
+        # values represent a map from measurements name to measurements
+        # definition. The values follow the expected `DictProvider` schema.
+        "10RandomIntsSumExercise": {
+            "ints": { # Measurement name.
+                "unit": "none", # We do not have a specific measurement unit.
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/framework/statistics/__init__.py:77]: It shouldn't be optional in the majority of cases. I've documented the `DictProvider` schema definition in this aspect. I agree, this example proves that mentioning the unit might be redundant sometimes. However, there is also a benefit for redudancy, as in adding more clarity to what the test does, by stating clearly how the measurement is defined and used within the test. Let me know if you consider we might take another course of action at this moment. I do not believe that adding support for optional `unit` field would bring much value at this time.

File: tests/framework/statistics/__init__.py [hunk 12]
```diff
@@ -1,12 +1,316 @@
...
+                    }
+                ]
+            }
+        },
+        "10thRandomIntExercise": {
+            "int": {
+                "unit": "none", # We do not have a specific measurement unit.
+                "statistics": [
+                    {
+                        "function": "ValuePlaceholder",
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/framework/statistics/__init__.py:97]: Done.

File: tests/framework/statistics/consumer.py [hunk 13]
```diff
@@ -4,56 +4,62 @@
...
+        # Final statistics.
         self._statistics = dict()
-        self._iteration = 0
 
     @abstractmethod
     def ingest(self, iteration: int, raw_data: Any):
         """Abstract method for ingesting the raw result."""
 
-    def consume_stat(self, st_name: str, ms_name: str, value: Any):
+    def consume_stat(self, st_name: str, ms_name: str, value: Number):
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/framework/statistics/consumer.py:50]: The abstraction would bring value as long as experimenting with a new test and making it ready to run becomes faster/clearer than what we have now. In regard to making it harder to grasp, indeed, we must avoid it if the value added/problem resolved is not significant. I would postpone this decision for when we propose a RFC.

File: tests/framework/statistics/criteria.py (hunk 9 above)
Comment (iulianbarbu) [reply tests/framework/statistics/criteria.py]: Done.

File: tests/framework/statistics/function.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/framework/statistics/function.py:42]: Added documentation for this function. Please take a look.

File: tests/framework/statistics/function.py [hunk 14]
```diff
@@ -3,155 +3,163 @@
...
-        if len(self.results) is 1:
-            return self.results[0]
-        return stdev(self.results)
-
-    @classmethod
-    def name(cls) -> str:
-        """Return an identifier for `Stddev`."""
-        return "stddev"
+        if len(result) is 1:
+            return result[0]
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/framework/statistics/function.py]: Yes. Good catch. Done.

File: tests/framework/statistics/metadata.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        ```
+        """
+        super().__init__()
+        for ms_name in measurements:
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/framework/statistics/metadata.py:108]: This would mean naming the logic that initialize the measurements dict, from the constructor, in a way which will indicate about what it does. I personally do not feel that a dedicated abstract method should be added for this case at this specific moment and I would wait to see if such interface is really needed, for other implementations as well.

File: tests/framework/statistics/types.py [hunk 16]
```diff
@@ -26,147 +15,51 @@ class MeasurementDef:
...
+    statistics: List['StatisticDef']
 
     @classmethod
-    def cpu_utilization_vmm(cls):
-        """Return vmm cpu utilization measurement definition."""
-        return MeasurementDef(
-            DefaultMeasurement.CPU_UTILIZATION_VMM.name.lower(),
-            "percentage"
-        )
+    def create_measurement(cls,
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/framework/statistics/types.py:21]: Unfortunately, dataclasses have already a predefined constructor [1], mapped on the fields declared above, so I can not provide a constructor with similar signature, since python classes do not accept multiple constructors.

[1] https://docs.python.org/3/library/dataclasses.html

File: tests/framework/utils.py [hunk 17]
```diff
@@ -216,6 +216,51 @@ def stopped(self):
...
+        """Initialize the dict query."""
+        self._inner = d
+
+    def get(self, keys_path: str, default=None):
+        """Retrieve value corresponding to the key path."""
+        keys = keys_path.strip().split("/")
+        if len(keys) < 1:
+            return default
+
+        result = self._inner.get(keys[0])
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/framework/utils.py]: Fixed it.

File: tests/integration_tests/performance/test_block_performance.py [hunk 18]
```diff
@@ -337,33 +208,34 @@ def read_values(cons, numjobs, env_id, mode, bs, measurement):
...
+            if len(values[measurement_id][idx]) != numjobs:
+                continue
 
-                value = sum(values[measurement_id][idx])
-                if DEBUG:
-                    cons.consume_custom(measurement_id, value)
-                cons.consume_measurement(measurement_id, value)
+            value = sum(values[measurement_id][idx])
+            if DEBUG:
+                cons.consume_custom(measurement_id, value)
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply tests/integration_tests/performance/test_block_performance.py:220]: Done.

File: tests/integration_tests/performance/test_vsock_throughput.py (hunk 8 above)
Comment (iulianbarbu) [reply tests/integration_tests/performance/test_vsock_throughput.py:344]: Done.

**Review by iulianbarbu — COMMENTED** [review]
@alindima, I've addressed your new feedback. I want to test first if performance tests run till the end on my private pipeline before moving on with merging the PR, if there is no other feedback. I'll let you know what is the outcome of my test runs shortly.

File: tests/framework/statistics/__init__.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (alindima) [reply tests/framework/statistics/__init__.py:11]: doesn't seem to be done yet

File: tests/framework/statistics/consumer.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (alindima) [reply tests/framework/statistics/consumer.py:50]: right, let's leave things as they are for now

File: tests/framework/statistics/metadata.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (alindima) [reply tests/framework/statistics/metadata.py:108]: I'm fine keeping it this way for now

**Review by alindima — APPROVED** [review]

**Comment by alindima** [comment]
great work 🚀

**Review by alindima — APPROVED** [review]

**Review by gc-plp — APPROVED** [review]
